Flatten artifacts on load

alanmcruickshank opened the pull request:
This is a draft solution to #62, #45, #20 & #78.

Results are flattened on upload into three tables:
- one for run result files (same granularity as artifacts)
- one for the flattened results together
- one for the flattened manifest elements (with sources, exposures compressed in with other nodes)

This upload is facilitated by a new `V2` upload macro, and then the results of both this approach and the old `V1` upload method are unioned together to give teams flexibility on which they use.

This should bypass field size limits (because we use `insert into`), and preloads the compute load of flattening large JSON blobs so allow larger projects to work well.

alanmcruickshank committed:
Add run_results query and schema

Manifest nodes and create table queries

Handle stages better

adjust insert into

Swap all from COPY TO to INSERT

Swtich names for artifact tables.

update sources file

Fix manifest node reference

correct the manifest time

Actually log the queries

Add sources and exposures

typo on insert queries

correct the schema for sources and exposures

fix wrong reference

alanmcruickshank committed:
Add instructions in README

alanmcruickshank committed:
Union in the V2 manifest elements

alanmcruickshank committed:
Merge remote-tracking branch 'origin/main' into ac/materialise_on_load

NiallRees commented on README.md:
```diff
@@ -46,32 +51,57 @@ Note that the model materializations are defined in this package's `dbt_project.
+   to your production run to load files into your table.
+   
+   **V2 Macro**: Use the `upload_artifacts_v2` macro ([source](macros/upload_artifacts.sql)). You only
+   need to run the macro after `run`, `test`, `seed` or `build` operations.
```
```suggestion
   need to run the macro after `build`, `snapshot`, `run`, `test`, `seed` or `build` operations.
```

~~The new approach means we just support the build command automatically right?~~ Can we just remove this filter now that we're filtering by resource type in each model? https://github.com/brooklyn-data/dbt_artifacts/blob/main/models/staging/stg_dbt__run_results.sql#L20

alanmcruickshank replied:
Yep.... and you put `build` in the list twice 🤣

NiallRees commented on macros/create_artifact_resources.sql:
```diff
@@ -9,21 +16,132 @@ create stage if not exists {{ src_dbt_artifacts }}
 
+{% set create_new_stage_query %}
+create stage if not exists {{ artifact_stage }}
+file_format = ( type =  json );
```
```suggestion
file_format = (type = json);
```

NiallRees commented on macros/create_artifact_resources.sql:
```diff
@@ -9,21 +16,132 @@ create stage if not exists {{ src_dbt_artifacts }}
+    command_invocation_id string,
+    dbt_cloud_run_id int,
+    artifact_run_id string,
+    artifact_generated_at timestamp_ntz,
```
Looks like all dbt artifact timestamps are in UTC: https://github.com/brooklyn-data/dbt_artifacts/blob/main/integration_test_project/artifacts/schema_v4/build/run_results.json#L5
```suggestion
    artifact_generated_at timestamp_tz,
```

NiallRees commented on macros/create_artifact_resources.sql:
```diff
@@ -1,6 +1,13 @@
+{% set artifact_stage = var('dbt_artifacts_stage', 'dbt_artifacts_stage') %}
+
+{% set src_results = source('dbt_artifacts', 'dbt_run_results') %}
+{% set src_result_nodes = source('dbt_artifacts', 'dbt_run_result_nodes') %}
```
```suggestion
{% set src_run_results = source('dbt_artifacts', 'dbt_run_results') %}
{% set src_run_results_nodes = source('dbt_artifacts', 'dbt_run_results_nodes') %}
```

Just to keep the names aligned with the artifacts exactly

NiallRees commented on macros/create_artifact_resources.sql:
```diff
@@ -9,21 +16,132 @@ create stage if not exists {{ src_dbt_artifacts }}
+{% do log("Creating Old Stage: " ~ create_stage_query, info=True) %}
 {% do run_query(create_stage_query) %}
 
+{% do log("Creating New Stage: " ~ create_new_stage_query, info=True) %}
```
Let's version these like you've done elsewhere

NiallRees commented on macros/upload_artifacts_v2.sql:
```diff
@@ -0,0 +1,218 @@
+        
+        select
+            metadata:invocation_id::string as command_invocation_id,
+            metadata:env:DBT_CLOUD_RUN_ID::int as dbt_cloud_run_id,
```
```suggestion
            metadata:env:dbt_cloud_run_id::int as dbt_cloud_run_id,
```

alanmcruickshank replied:
I think this is actually case sensitive - might be remembering wrong but I'll add a comment to that effect.

NiallRees commented on macros/upload_artifacts_v2.sql:
```diff
@@ -0,0 +1,218 @@
+                run_results.$1:metadata as metadata,
+                run_results.$1:args as args,
+                run_results.$1:elapsed_time::float as elapsed_time
+            from  @{{ artifact_stage }} as run_results
```
```suggestion
            from @{{ artifact_stage }} as run_results
```

NiallRees commented on macros/upload_artifacts_v2.sql:
```diff
@@ -0,0 +1,218 @@
+{% endset %}
+
+{% set manifest_nodes_query %}
+    begin;
```
```suggestion
    begin;
```

Not sure that the begin; and commit; are necessary - the run_query macro handles all that to my knowledge

NiallRees commented on models/staging/sources.yml:
```diff
@@ -18,3 +19,89 @@ sources:
+        identifier: "{{ var('dbt_artifacts_manifest_exposures_table', 'dbt_run_manifest_exposures') }}"
+        description: |
+          The source table containing the loaded and flattened exposures from manifest.json loaded artifacts. This belongs
+          to the _new style_ upload. See the README for more info.
```
Let's avoid relative versioning and go for v1 and v2 like elsewhere

alanmcruickshank committed:
Merge remote-tracking branch 'origin/main' into ac/materialise_on_load

alanmcruickshank committed:
Move flatten logic to macro

alanmcruickshank committed:
Shift upload macro to the new flatten macros

alanmcruickshank committed:
artifact run id macro

alanmcruickshank committed:
Update artifact creation with new schema

alanmcruickshank committed:
union in the V1 and V2 uploads

alanmcruickshank committed:
Migrate env values over to depend on the run results staging model

alanmcruickshank committed:
Misc. Fixes from production testing

alanmcruickshank committed:
Remove build type filters

alanmcruickshank committed:
Switch to Timezone aware timestamps

alanmcruickshank committed:
consistent results naming

alanmcruickshank committed:
consitent v1/v2 naming

alanmcruickshank committed:
comment on case sensitivity

alanmcruickshank committed:
Remove begin/commit

alanmcruickshank committed:
Update sources file

alanmcruickshank committed:
Fix source reference

alanmcruickshank committed:
fix table name ref

alanmcruickshank committed:
update plurals

alanmcruickshank committed:
force timestamps

alanmcruickshank committed:
more timestamp forcing

alanmcruickshank committed:
remove unnecessary whitespace

NiallRees commented on models/staging/sources.yml:
```diff
@@ -18,3 +19,62 @@ sources:
+            description: Unique id for the node, in the form of exposure.[package_name].[exposure_name]
+          - name: node_json
+            description: Raw JSON structure of the manifest node.
+  
```
```suggestion
```

alanmcruickshank committed:
Update models/staging/sources.yml

Co-authored-by: Niall Woodward <niall@niallrees.com>

alanmcruickshank committed:
Integrate a replace option into creation macro for testing.

alanmcruickshank committed:
Add a V2 integration test setup

alanmcruickshank commented:
@NiallRees - Check out what I've tried with the integration suite.

- I've added an optional `replace` argument for the create macro so we can get fresh artifact tables without duplicating the code.
- I've made a separate integration suite for the V2 upload. In theory we should have the same result whether we use the V1 or V2 upload. I don't know if we have any data tests in place - but if we did then they should work in either case.

NiallRees commented on tox.ini:
```diff
@@ -103,6 +104,21 @@ commands =
+    dbt deps
+    # Setting the replace arg to tru so that we get a clean set of tabels on each run.
+    dbt run-operation create_artifact_resources --args '\{replace: true\}'
+    dbt run-operation upload_dbt_artifacts_v2
```
Can we test uploading all of the files in the `schema_v4` directory here like the above integration tests do?

alanmcruickshank replied:
👍

NiallRees commented on macros/create_artifact_resources.sql:
```diff
@@ -1,29 +1,115 @@
-{% set create_table_query %}
-create table if not exists {{ src_dbt_artifacts }} (
+{% set create_v1_table_query %}
+create {% if replace %} or replace {% endif %} table if not exists {{ src_dbt_artifacts }} (
```
Unfortunately `or replace` and `if not exists` can't be used in the same statement: https://docs.snowflake.com/en/sql-reference/sql/create.html#general-usage-notes

NiallRees replied:
For the purposes of these automated integration tests the `replace` doesn't matter as each runs in a separate schema named by the commit hash. Appreciate it's useful for local testing though.

alanmcruickshank replied:
ahhhhh - understand. I thought it was _necessary_ for the CI tests! In that case I'm just going to revert the `replace` changes. I think it's almost a _dangerous_ feature if you don't need it.

alanmcruickshank committed:
Revert the `or replace` feature.

alanmcruickshank committed:
add prefix into upload macro

alanmcruickshank committed:
Set the prefix to test all the different artifacts

alanmcruickshank commented:
Ok I think this is ready for another review:
- I removed the `or replace` option entirely. I think it's a dangerous feature unless we need it for CI testing.
- I've added the `prefix` option into the new upload macro so we can test all the new artifacts.

I intentionally haven't added the `sources` file as one of the things we test - because none of the flatten code handles it. I don't know if this is an issue because I can't see any code in the rest of the package which actually _handles_ a sources file. I assume that would be tackled as a totally separate issue?

alanmcruickshank committed:
clean the schema before running the suite

alanmcruickshank committed:
Implement basic (and probably failing) tests

alanmcruickshank committed:
it's probably seven

alanmcruickshank committed:
Add real results

alanmcruickshank committed:
Provide args

NiallRees commented on .github/workflows/ci_test_package.yml:
```diff
@@ -37,5 +37,8 @@ jobs:
         run: tox -e integration
+      
+      - name: Run V2 Tests
+        run: tox -e integration_v2
```
```suggestion
      - name: Run Tests
        run: tox -e integration
```

NiallRees commented on .github/workflows/main_test_models.yml:
```diff
@@ -33,5 +33,8 @@ jobs:
         run: tox -e integration
+      
+      - name: Run V2 Tests
+        run: tox -e integration_v2
```
```suggestion
      - name: Run Tests
        run: tox -e integration
```
```

NiallRees committed:
Apply suggestions from code review

NiallRees commented on README.md:
```diff
@@ -46,32 +51,57 @@ Note that the model materializations are defined in this package's `dbt_project.
 3. Run `dbt deps`.
 
 ## Generating the source table
```
```suggestion
## Uploading the artifacts
```

NiallRees commented on README.md:
```diff
@@ -46,32 +51,57 @@ Note that the model materializations are defined in this package's `dbt_project.
-This package requires that the source data exists in a table in Snowflake.
-
-### Option 1: Loading local files
+This package requires that the source data exists in Snowflake. There are two supported ways of doing this:
```
```suggestion
This package uploads the artifact files into Snowflake. There are two supported ways of doing this:
```

NiallRees commented on README.md:
```diff
@@ -46,32 +51,57 @@ Note that the model materializations are defined in this package's `dbt_project.
+  _"local file"_ method. Loading via cloud storage is also _possible_ but we recommend users
+  copy the method used in `upload_artifacts_v2.sql` to create their own approach.
+- The _V1_ or _legacy_ option, which uploads the files unprocessed. This affords much more flexibility
+  in their use, but the is subject to field size limits and higher compute loads to reprocess the
```
```suggestion
  in their use, but is subject to field size limits and higher compute loads to reprocess the
```

NiallRees committed:
Apply suggestions from code review

NiallRees reviewed (approved):
Great job. Tests too!
